while testing, when a `DynaKube` updated with a `namespaceSelector` that deselects a previously-injected namespace, the validating webhook panics with a nil pointer dereference.

my steps:
- `kubectl create namespace test-ns`
- `kubectl label namespace test-ns env=test`
- apply a `DynaKube` that selects namespaces with label `env=test`
- wait until secrets have been replicated
- apply patched `DynaKube` with `env=prod` label selector
- get ```Error from server: admission webhook "v1beta6.dynakube.webhook.dynatrace.com" denied the request:
panic: runtime error: invalid memory address or nil pointer dereference [recovered]```

--- a/pkg/injection/namespace/mapper/dynakubes.go
+++ b/pkg/injection/namespace/mapper/dynakubes.go
@@ -31,6 +31,8 @@ type DynakubeMapper struct {
 	matchedOANamespaces   []string
 	matchedMENamespaces   []string
 	matchedOTLPNamespaces []string
+
+	deselectedNamespaces []string
 }
 
 func NewDynakubeMapper(ctx context.Context, clt client.Client, apiReader client.Reader, operatorNs string, dk *dynakube.DynaKube) DynakubeMapper {
@@ -46,6 +48,7 @@ func NewDynakubeMapper(ctx context.Context, clt client.Client, apiReader client.
 		matchedOANamespaces:   []string{},
 		matchedMENamespaces:   []string{},
 		matchedOTLPNamespaces: []string{},
+		deselectedNamespaces:  []string{},
 	}
 }
 
@@ -60,6 +63,12 @@ func (dm *DynakubeMapper) MapFromDynakube() error {
 
 	for _, ns := range modifiedNs {
 		if err := dm.client.Update(dm.ctx, ns); err != nil {
+			return err
+		}
+	}
+
+	for _, nsName := range dm.deselectedNamespaces {
+		if err := dm.deleteReplicatedSecrets(nsName); err != nil {
 			return err
 		}
 	}
@@ -181,9 +190,7 @@ func (dm *DynakubeMapper) mapNamespace(namespace *corev1.Namespace, dkList *dyna
 	}
 
 	if previouslyInjected && !result.IsAny() {
-		if err := dm.deleteReplicatedSecrets(namespace.Name); err != nil {
-			return false, err
-		}
+		dm.deselectedNamespaces = append(dm.deselectedNamespaces, namespace.Name)
 	}
 
 	return updated, nil

--- a/pkg/injection/namespace/mapper/dynakubes_test.go
+++ b/pkg/injection/namespace/mapper/dynakubes_test.go
@@ -179,6 +179,47 @@ func TestMapFromDynakube(t *testing.T) {
 	})
 }
 
+func TestMatchingNamespaces(t *testing.T) {
+	t.Run("validating webhook run on deselected ns => replicated secrets not deleted", func(t *testing.T) {
+		matchingLabels := map[string]string{"foo": "bar"}
+		dk := createDynakubeWithAppInject("dk-test", convertToLabelSelector(matchingLabels))
+
+		namespace := createNamespace(
+			"definitely-not-selected",
+			map[string]string{
+				dtwebhook.InjectionInstanceLabel: dk.Name,
+			},
+		)
+
+		clt := fake.NewClient(dk, namespace)
+		ctx := t.Context()
+
+		secretNames := []string{
+			consts.BootstrapperInitSecretName,
+			consts.BootstrapperInitCertsSecretName,
+			consts.OTLPExporterSecretName,
+			consts.OTLPExporterCertsSecretName,
+		}
+		for _, secretName := range secretNames {
+			createSecret(t, clt, secretName, namespace.Name)
+		}
+
+		// pass nil write client, just like validating webhooks
+		dm := NewDynakubeMapper(ctx, nil, clt, "dynatrace", dk)
+
+		require.NotPanics(t, func() {
+			_, err := dm.MatchingNamespaces()
+			require.NoError(t, err)
+		})
+
+		for _, secretName := range secretNames {
+			var secret corev1.Secret
+			err := clt.Get(ctx, types.NamespacedName{Name: secretName, Namespace: namespace.Name}, &secret)
+			require.NoErrorf(t, err, "secret %s must not be deleted when validating webhooks are run", secretName)
+		}
+	})
+}
+
 func TestUnmapFromDynaKube(t *testing.T) {
 	dk := createDynakubeWithAppInject("dk", metav1.LabelSelector{})
 	labels := map[string]string{

